Team417_PowerPlay/src/main/java/org/firstinspires/ftc/team417_PowerPlay/BaseTeleOp.java

```diff
@@ -1,44 +1,159 @@
 package org.firstinspires.ftc.team417_PowerPlay;
 
+import com.qualcomm.robotcore.hardware.DcMotor;
+import com.qualcomm.robotcore.util.ElapsedTime;
+import com.qualcomm.robotcore.util.Range;
+
 abstract public class BaseTeleOp extends BaseOpMode {
-    boolean grabberOpen = false;
+    boolean grabberClosed = false;
     double armPower = 0.0;
+    int armLevel = 0;
+    int armEncoderPosition = 0;
+
+    boolean armManual = false;
+
+    ElapsedTime time;
+    private static double BUMPER_TIMEOUT = 0.3; // in seconds
+
+    private static double LOWER_ARM_MANUAL_POWER = -0.2;
+    private static double LOWER_ARM_AUTO_POWER = -0.4;
+    private static double RAISE_ARM_MANUAL_POWER = 1.0 / 100.0;
+    private static double RAISE_ARM_AUTO_POWER = 1.0 / 800.0;
 
-    private static double DRIVING_SPEED = 0.5;
+    private static double DRIVING_SPEED = 0.7;
+
+    public void initializeTeleOp() {
+        initializeHardware();
+        time = new ElapsedTime();
+        time.reset();
+
+        grabberServo.setPosition(GRABBER_OPEN);
+
+    }
 
     public void driveUsingControllers() {
         double x = gamepad1.left_stick_x * DRIVING_SPEED;
         double y = -gamepad1.left_stick_y * DRIVING_SPEED;
         double turning = gamepad1.right_stick_x * DRIVING_SPEED;
+        y *= 1 - (0.5 * gamepad1.right_trigger);
+        x *= 1 - (0.5 * gamepad1.right_trigger);
+        turning *= 1 - (0.5 * gamepad1.right_trigger);
+        mecanumDrive(x, y, turning);
+    }
+
+    public double stickCurve(double x, double a, double d) {
```

Comment:
This function should have a javadoc comment since it's not very self-explanatory. What is the curve you are using? How was it determined?

Reply:
You need a bit more here. What shape is the curve? How did you mathematically determine that this was the appropriate curve?

Reply:
I added more explanation :)